### DISCUSSION TITLE
Enable classic MIDI input

### DISCUSSION THREAD

**Discussion by luisfcorreia** [probonopd/MiniDexed#104] [General]
It would be nice to be able to send classic MIDI data through RX pin

It only needs a simple optocoupler circuit for acccepting notes and other data.

**Comment by luisfcorreia** [comment]
maybe already present, will test it soon

(https://github.com/probonopd/MiniDexed/blob/main/src/serialmididevice.cpp)

**Comment by probonopd** [comment]
Welcome @luisfcorreia.

Yes, it is already present software-wise. You still need to build the optocoupler circuit. Please let us know which exact circuit you are using if you succeed! (I had played with this very briefly but never got it to work electrically.)

GPIO 15 (UART RX) is the pin to be used on the Raspberry Pi.

![image](https://user-images.githubusercontent.com/2480569/163724149-d8202d75-b404-46a6-9085-c1f05a9f4a3e.png)

**Reply by luisfcorreia** [reply]
Sure, will do it and provide usable data

**Reply by probonopd** [reply]
This had been my starting point, but I gave up before I had it working (mainly because my Serial-MIDI keyboard has no velocity sensitivity which makes it a toy and I lost interest in it):

![image](https://user-images.githubusercontent.com/2480569/152688743-9f07ca3f-f851-4c32-bbf1-0aaab8bea06c.png)

Image Source: https://hackaday.com/tag/6n137/

What I have already figured out is that the circuit is picky about the optocoupler being fast enough. One can't just use any random type.

Connect the wire labeled "RX" to GPIO15 on the Raspberry Pi. This is a SoC number, not the position on the header.

**Reply by luisfcorreia** [reply]
well, the resistor must be of lower value, something along the 4.7k or lower, as this is for 3.3V

**Reply by probonopd** [reply]
This could well be it! 4.7k instead of the 10k resistor?

**Reply by arsamus** [reply]
I've build a similar circuit but with 6n138 optocoupler and I've use a 2k2 as pullup resistor instead of  10k. I've tested on a Raspberry 1B+, works ok if a press one note at a time, but I can't play chords. It seems that miss some note events (Note On or Note Off). Some notes continue playing forever. Maybe it is a buffer issue.

**Reply by Kirtai** [reply]
MIDI components for 3.3V circuits are different from the 5V circuit you see everywhere.

**Reply by craigyjp** [reply]
No they are not, pretty much the same components are used just with different values of resistors for 5v or 3.3v circuits. I have built 5v and 3.3v circuits for many designs and they are pretty much the same.

**Reply by Kirtai** [reply]
Yes, the circuit is the same but the optocoupler should be a H11L1 or PC900. Though many manufacturer's models of the 6N137 do handle 3.3v it's only guaranteed for 5V.

**Reply by luisfcorreia** [reply]
my circuit with a 6n139, as described in the WIKI is working with 3.3V

could we please stop this discussion?

**Reply by probonopd** [reply]
Thanks for adding the solution to the wiki @luisfcorreia. Will build one as soon as I can get the parts.

**Reply by craigyjp** [reply]
I agree, everyone has success with different optocouplers, I prefer 4n35 myself, but have used 6n137 and 6n138 at 5v and 3.3v

**Comment by eLuSiVeMiTe** [comment]
Thought I'd piggyback here. 
I'm planning on using a old novation SL mk2 with this as the project evolves (will make an excellent programmer) the keyboard also acts as a midi interface. Is there anyway to get midi din from the interface into dexed. I could slot the SL and pi into my rig as a hardware synth then.

**Reply by probonopd** [reply]
Hello @eLuSiVeMiTe. If I understand you correctly, you'd like to have a DIN-style MIDI IN on MiniDexed. Correct? This is what we are talking about here.

Another option of course is to use a USB-to-DIN-Midi adapter.

**Reply by eLuSiVeMiTe** [reply]
Evening mate. Great project. Got it running earlier but waiting for DAC/encoder)screen to dig in deeper. 

 Usb to din.  My midi controller is also a midi interface. So I could use the controller to program and play live and also send midi through it from my hardware sequencers all via usb. 

Though if is not possible il go the midi din route mentioned above

**Reply by probonopd** [reply]
So you want MiniDexed to send MIDI received via USB over serial out? Would your need be satisfied by https://github.com/probonopd/MiniDexed/issues/105?

**Comment by eLuSiVeMiTe** [comment]
Sorry I'm not explaining myself well. 


MPC sequencer >>>midi din>>> midi controller>>>usb midi >>>> dexed 

So dexed would respond to 2 midi inputs at the same time. From both the MPC and the midi controller 



The midi controller usually displays in a DAW as 2 midi inputs. 

Hope that makes more sense.

**Reply by probonopd** [reply]
That should already work, unless I am misunderstanding something. Try it out!

**Comment by eLuSiVeMiTe** [comment]
Yup. It does. Perfect. Can use the controller as a dedicated keyboard for dexed and to push midi from my rig too. :)
https://youtu.be/eUM1Q146qC8
